### PR TITLE
AMSEOS Wrapper

### PR DESCRIPTION
**Pull request by tomstitt**
- Move `AMSConfig`+`AMSExecutor` and `AMS*Execute` to an `AMSEOS` that wraps an `EOS`
   - Cleanup main
- Move `Eval(const int length, const FPType **inputs, FPType **outputs)` to base `EOS` and remove implementations from derived classes

### CODE REVIEW

File: examples/main.cpp [hunk 1]
```diff
@@ -361,26 +333,45 @@ int run(const char *device_name,
...
+    if (eos_name == std::string("ideal_gas")) {
+      base = new IdealGas<TypeValue>(1.6, 1.4);
+    } else if (eos_name == std::string("constant_host")) {
+      base = new ConstantEOSOnHost<TypeValue>(physics_host_alloc.c_str(), 1.0);
+    } else {
+      std::cerr << "unknown eos `" << eos_name << "'" << std::endl;
+      return 1;
+    }
+#ifdef USE_AMS
+    if (/*use_ams*/true) {
```
Comment (koparasy) [inline examples/main.cpp]: Do we need this if statement?

File: examples/main.cpp [hunk 2]
```diff
@@ -361,26 +333,45 @@ int run(const char *device_name,
...
+      eoses[mat_idx] = new AMSEOS<TypeValue>(base,
```
Comment (koparasy) [inline examples/main.cpp:357]: Are we releasing this memory somewhere?

**Review by koparasy — CHANGES_REQUESTED** [review]
Did you clang-format this?

File: examples/main.cpp (hunk 1 above)
Comment (tomstitt) [reply examples/main.cpp]: I changed this so we have it `true` with `USE_AMS` is defined, `false` otherwise

File: examples/main.cpp (hunk 2 above)
Comment (tomstitt) [reply examples/main.cpp:357]: Yes, `AMSEOS` owns `base` and will `delete` it and at the end of `main` we remove everything from `eoses`

**Comment by tomstitt** [comment]
> Did you clang-format this?

done!

**Review by koparasy — APPROVED** [review]

**Pull request merged**